samples/Calling/src/app/views/CallScreen.tsx

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import React, { useEffect, useRef, useState } from 'react';
+import { Spinner } from '@fluentui/react';
+import { GroupCallLocator, TeamsMeetingLinkLocator } from '@azure/communication-calling';
+import { CallAdapter, CallComposite, createAzureCommunicationCallAdapter } from '@azure/communication-react';
+import { CommunicationUserIdentifier } from '@azure/communication-common';
+import { refreshTokenAsync } from '../utils/refreshToken';
+import { useSwitchableFluentTheme } from '../theming/SwitchableFluentThemeProvider';
+
+export interface CallScreenProps {
+  token: string;
+  userId: CommunicationUserIdentifier;
+  callLocator: GroupCallLocator | TeamsMeetingLinkLocator;
+  displayName: string;
+  onCallEnded: () => void;
+  onCallError: (e: Error) => void;
+}
+
+export const CallScreen = (props: CallScreenProps): JSX.Element => {
+  const { token, userId, callLocator, displayName, onCallEnded, onCallError } = props;
+  const [adapter, setAdapter] = useState<CallAdapter>();
+  const adapterRef = useRef<CallAdapter>();
+  const { currentTheme } = useSwitchableFluentTheme();
+
+  useEffect(() => {
+    (async () => {
+      console.log('creating adapter');
+      const adapter = await createAzureCommunicationCallAdapter(
+        userId,
+        token,
+        callLocator,
+        displayName,
+        refreshTokenAsync(userId.communicationUserId)
+      );
+      adapter.on('callEnded', () => {
+        onCallEnded();
+      });
+      adapter.on('error', (e) => {
+        console.error(e);
+        onCallError(e);
+      });
+      setAdapter(adapter);
+      adapterRef.current = adapter;
+    })();
+
+    return () => {
+      adapterRef?.current?.dispose();
```

Comment:
Is there reason to add adapterRef here? If that is just for unmount, add the adapter to dependency array should work

Reply:
It was to ensure the dispose was called (otherwise I got a 'Call Agent already exists' error on rejoining a call)

Reply:
Yea, that is for unmount right? If you put adapter into dependency array and unmount directly using adapter.?dispose() that should work

Reply:
If no adapter in dependency array, adapter in unmount function will be always 'undefined'

Reply:
But you might need to add an if-else to adapter creation logic above

Reply:
The adapter is set asyncly in the useEffect, so my thought was the return will capture the original adapter reference (i.e. undefined) in its closure. But let me test that

Reply:
The adapter is set asyncly in the useEffect, so my thought was the return will capture the original adapter reference (i.e. undefined) in its closure.
If we put the adapter in the dep array that useEffect will be called twice - which I guess we can do a `if(!adapter) createAdapter()` but isn't this what useRef is for?